shapes.sc

```diff
@@ -288,6 +290,106 @@ line_sight(p1, material, length) -> (
 	line(pos1, end_point, material, 1);
 );
 
+tube_circle(p1,p2,p3,material,width) ->(
+  v1 = p2-p1;
+  v3 = p3-p1;
+  vmajor = v3 - __proj(v3,v1);
+  rmajor = __norm(vmajor);
+  rminor = rmajor;
+  magv1=__norm(v1);
+  cmin=l(0,0,0);
+  cmax=l(0,0,0);
+  loop(3,
+    
+    dn = sin(acos(get(v1,_)/magv1))*(max(rmajor,rminor)+width);
+    print('dn = '+str(dn));
+    if( get(v1,_)>=0,
+      put(cmin, _, floor(get(p1,_)-dn));
+      put(cmax, _, ceil(get(p2,_)+dn));
+    );
+    if( get(v1,_)<0,
+      put(cmin, _, floor(get(p2, _) - dn));
+      put(cmax, _, ceil(get(p1, _) + dn));
+    );
+  );
+  print(str(cmin));
+  print(str(cmax));
+  hminor = __normalize(__cross_prod(vmajor,v1));
+  hmajor = __normalize(vmajor);
+  //print(str(hminor)+','+str(hmajor));
+  volume(cmin,cmax,
+    tv=pos(_)-p1;
+    tp=pos(_);
+    magProjMajor = __dot_prod(tv,hmajor);
+    magProjMinor = __dot_prod(tv,hminor);
+    
+    //this long if statement is the eqivlent of the check function from the python ver.
+    //inside=((magProjMajor^2)/(rmajor^2)+(magProjMinor^2)/(rminor^2))>=1;
+    //outside=((magProjMajor)^2/(rmajor+width)^2+(magProjMinor)^2/(rminor+width)^2)<=1;
+    //print(str((magProjMajor)^2/(rmajor+width)^2)+'+'+str(magProjMinor));
+    //print(str(inside)+','+str(outside)+':'+str(tp));
+    started = (__dot_prod(tv,v1))>=0;
+    notended = (__dot_prod(tp-p2, -1*v1))>=0;
+    if( (((magProjMajor^2)/(rmajor^2)+(magProjMinor^2)/(rminor^2))>=1 &&
+          ((magProjMajor)^2/(rmajor+width)^2+(magProjMinor)^2/(rminor+width)^2)<=1 
+          && (__dot_prod(tv,v1))>=0 && (__dot_prod(tp-p2, -1*v1))>=0 ),

```

Comment:
The function is a bit slow right now. Maybe saving some calculation time by calculating some constants before hand makes it faster. Stuff like `oorms = 1/rmajor^2 //one over r major squared` (you are free to think of better names. See if doing that improves the execution times. Same goes for the squares of `magProjMayor` and `-Minor`, btw.